Arm/BitVec.lean

```diff
@@ -813,4 +813,84 @@ macro_rules
 
 -- #print test2
 
+/-! ### Equation Lemmas -/
+
+theorem gt_def {x y : BitVec w} : (x > y) = (x.toNat > y.toNat) := rfl
+theorem ge_def {x y : BitVec w} : (x ≥ y) = (x.toNat ≥ y.toNat) := rfl
+
+theorem neq_of_toNat_neq {x y : BitVec w} (hx : x.toNat ≠ y.toNat) : x ≠ y := by
+  intros h
+  simp [h] at hx
+
+/-! ### (Non) Overflow Rewriting Lemmas
+
+@bollu: To be upstreamed.
+-/
+
+private theorem Nat.sub_add_eq_add_sub_of_le_of_le {a b c : Nat} (hab : b ≤ a) (hbc : b ≤ c) :
+    a - b + c = a + (c - b) := by
+  omega
+
+/- TODO: upstream. -/
+theorem toNat_sub_eq_toNat_sub_toNat_of_le {x y : BitVec w} (h : y ≤ x) :
+    (x - y).toNat = x.toNat - y.toNat := by
+  simp only [toNat_sub]
+  rw [BitVec.le_def] at h
+  by_cases h' : x.toNat = y.toNat
+  · rw [h', Nat.sub_self]
+    rw [Nat.sub_add_cancel (by omega), Nat.mod_self]
+  · rw [Nat.sub_add_eq_add_sub_of_le_of_le (by omega) (by omega), Nat.add_mod,
+      Nat.mod_self, Nat.zero_add, Nat.mod_mod, Nat.mod_eq_of_lt (by omega)]
+
+theorem neq_of_lt {x y : BitVec w₁} (h : x < y) : x ≠ y := by
+  rintro rfl
+  simp [BitVec.lt_def] at h
+
+theorem neq_of_gt {x y : BitVec w₁} (h : x > y) : x ≠ y := by
+  rintro rfl
+  simp [BitVec.lt_def] at h
+
+
+/-- adding bitvectors when there is no overflow. -/
+theorem toNat_add_eq_toNat_add_toNat {x y : BitVec w} (h : x.toNat + y.toNat < 2^w) :
+    (x + y).toNat = x.toNat + y.toNat := by
+  rw [BitVec.toNat_add, Nat.mod_eq_of_lt h]
+
+/-! ### Least Significant Byte -/
+
+/-- Definition to extract the `n`th least significant *Byte* from a bitvector. -/
+def extractLsByte (val : BitVec w₁) (base : Nat) : BitVec 8 :=
+  val.extractLsb ((base + 1)*8 - 1) (base * 8) |>.cast (by omega)
+
+theorem extractLsByte_def (val : BitVec w₁) (n : Nat) :
+    val.extractLsByte n = (val.extractLsb ((n + 1)*8 - 1) (n * 8) |>.cast (by omega)) := rfl
+
+@[simp]
+theorem getLsb_extractLsByte (val : BitVec w₁) :
+    ((BitVec.extractLsByte val n).getLsb i) =
+    (decide (i ≤ 7) && val.getLsb (n * 8 + i)) := by
+  simp [extractLsByte]
```

Comment:
Stick to `simp only` in LNSym as far as possible.

Reply:
Fixed.